edb/server/pgcon/rust_transport.py

```diff
@@ -0,0 +1,475 @@
+                          f"the provided hosts: {host_candidates}") from e
+
+
+def _set_tcp_keepalive(transport):
+    # TCP keepalive was initially added here for special cases where idle
+    # connections are dropped silently on GitHub Action running test suite
+    # against AWS RDS. We are keeping the TCP keepalive for generic
+    # Postgres connections as the kernel overhead is considered low, and
+    # in certain cases it does save us some reconnection time.
+    #
+    # In case of high-availability Postgres, TCP keepalive is necessary to
+    # disconnect from a failing master node, if no other failover information
+    # is available.
+    sock = transport.get_extra_info('socket')
+    sock.setsockopt(socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)
+
+    # TCP_KEEPIDLE: the time (in seconds) the connection needs to remain idle
+    # before TCP starts sending keepalive probes. This is socket.TCP_KEEPIDLE
+    # on Linux, and socket.TCP_KEEPALIVE on macOS from Python 3.10.
+    if hasattr(socket, 'TCP_KEEPIDLE'):
+        sock.setsockopt(socket.IPPROTO_TCP,
+                        socket.TCP_KEEPIDLE, TCP_KEEPIDLE)
+    if hasattr(socket, 'TCP_KEEPALIVE'):
+        sock.setsockopt(socket.IPPROTO_TCP,
+                        socket.TCP_KEEPALIVE, TCP_KEEPIDLE)
+
+    # TCP_KEEPINTVL: The time (in seconds) between individual keepalive probes.
+    if hasattr(socket, 'TCP_KEEPINTVL'):
+        sock.setsockopt(socket.IPPROTO_TCP,
+                        socket.TCP_KEEPINTVL, TCP_KEEPINTVL)
+
+    # TCP_KEEPCNT: The maximum number of keepalive probes TCP should send
+    # before dropping the connection.
+    if hasattr(socket, 'TCP_KEEPCNT'):
+        sock.setsockopt(socket.IPPROTO_TCP,
+                        socket.TCP_KEEPCNT, TCP_KEEPCNT)
+
+
+P = TypeVar('P', bound=asyncio.Protocol)
+
+
+async def create_postgres_connection(
+    dsn: str | ConnectionParams,
+    protocol_factory: Callable[[], P],
+    *,
+    source_description: Optional[str] = None,
+    state_change_callback: Optional[StateChangeCallback] = None
+) -> Tuple[PGRawConn, P]:
+    """
+    Open a PostgreSQL connection to the address specified by the DSN or
+    ConnectionParams, creating the user protocol from the protocol_factory.
+
+    This method establishes the connection asynchronously. When successful, it
+    returns a (PGRawConn, protocol) pair.
+    """
+    if isinstance(dsn, str):
+        dsn = ConnectionParams(dsn=dsn)
+    connect_timeout = dsn.connect_timeout
+    try:
+        state = PyConnectionState(
+            dsn._params,
+            "postgres",
+            str(get_pg_home_directory())
+        )
+    except Exception as e:
+        raise ValueError(e)
+    ready_future: asyncio.Future = asyncio.Future()
+    pg_state = PGState(
+        parameters={},
+        cancellation_key=None,
+        auth=None,
+        server_error=None,
+        ssl=False
+    )
+
+    user_protocol = protocol_factory()
```

Comment:
This may not be a problem now, but I'd defer the user protocol instantiation to when it's really needed (that also means multiple instantiations during retry), so that we don't get to the wrong case when the same user protocol instance gets multiple `connection_made()` calls in the future by mistake.

Reply:
With the latest changes, we only create this at the last moment